tools/python.js

```diff
@@ -0,0 +1,242 @@
+ * will make one tty for stdout and one for stderr, the `outstream` argument
+ * controls which one we make.
+ *
+ * Note that setting Module.stdin / Module.stdout / Module.stderr does not work
+ * because these cause `isatty(stdstream)` to return false. We want `isatty` to
+ * be true. (IMO this is an Emscripten issue, maybe someday we can fix it.)
+ */
+function make_tty_ops(outstream) {
+    let newline = false;
+    return {
+        // get_char has 3 particular return values:
+        // a.) the next character represented as an integer
+        // b.) undefined to signal that no data is currently available
+        // c.) null to signal an EOF
+        get_char(tty) {
+            if (newline) {
+                newline = false;
+                return null;
+            }
+            if (!tty.input.length) {
+                var result = null;
+                var BUFSIZE = 256;
+                var buf = Buffer.alloc(BUFSIZE);
+                // read synchronously at most BUFSIZE bytes from file descriptor 0 (stdin)
+                var bytesRead = fs.readSync(0, buf, 0, BUFSIZE, -1);
+                if (bytesRead === 0) {
+                    return null;
+                }
+                result = buf.slice(0, bytesRead);
+                tty.input = Array.from(result);
+            }
+            let res = tty.input.shift();
+            newline = true;
+            return res;
+        },
+        put_char(tty, val) {
+            outstream.write(Buffer.from([val]));
+        },
+        flush(tty) {},
+        fsync() {},
+    };
+}
+
+/**
+ * Fix standard streams by replacing them with ttys that work better via make_tty_ops.
+ */
+function setupStreams(FS, TTY) {
+    // Create and register devices
+    let mytty = FS.makedev(FS.createDevice.major++, 0);
+    let myttyerr = FS.makedev(FS.createDevice.major++, 0);
+    TTY.register(mytty, make_tty_ops(process.stdout));
+    TTY.register(myttyerr, make_tty_ops(process.stderr));
+    // Attach devices to files
+    FS.mkdev("/dev/mytty", mytty);
+    FS.mkdev("/dev/myttyerr", myttyerr);
+    // Replace /dev/stdxx with our custom devices
+    FS.unlink("/dev/stdin");
+    FS.unlink("/dev/stdout");
+    FS.unlink("/dev/stderr");
+    FS.symlink("/dev/mytty", "/dev/stdin");
+    FS.symlink("/dev/mytty", "/dev/stdout");
+    FS.symlink("/dev/myttyerr", "/dev/stderr");
+    // Refresh std streams so they use our new versions
+    FS.closeStream(0);
+    FS.closeStream(1);
+    FS.closeStream(2);
+    FS.open("/dev/stdin", 0);
+    FS.open("/dev/stdout", 1);
+    FS.open("/dev/stderr", 1);
+}
+
+/**
+ * Determine which native top level directories to mount into the Emscripten
+ * file system.
+ *
+ * This is a bit brittle, if the machine has a top level directory with certain
+ * names it is possible this could break. The most surprising one here is tmp, I
+ * am not sure why but if we link tmp then the process silently fails.
```

Comment:
Maybe because /tmp could be on a different filesystem, e.g on a tmpfs.

Why do we need to mount all the other dirs though? Wouldn't it be better / safer to mount only the current folder and its subdirs (or at least have an opt-in way to mount all of these).

A concern is that by now the users might be used that pyodide runs in a in memory FS,  including in node. Now if you give them the real FS, they might do things that break their system without realising they are not modifying the in memory FS.

Reply:
The goal of the command line runner is to be as close as we can manage to a drop-in replacement for ordinary python. This works very poorly with the sandboxed file system. For example, in `python path/to/somefile.py`, Python is going to look up the file to execute in the Emscripten file system and if it isn't in there then it will fail. The same is true with any files in arguments. If we are going to have a virtual environment, the virtual environment needs to be included. `pytest` has to look up the test files and whatever other stuff it needs. In my tests these files are not typically in the current folder or its subdirectories. The details of this are a bit different between running locally, vs CircleCI or github actions.

I also don't think this is very useful except for testing that functionality works with Pyodide. I think aside from that, this is worse in every way than a native Python. 

Maybe we could set up an environment variable to specify which directories to mount though.